File: doubly_linked_list/c/doubly_linked_list.c
#include "doubly_linked_list.h"
#include <stdio.h>

ListNode *list_search(DoublyLinkedList list, int value) {
  ListNode *curr = list.head;
  while (curr != NULL && curr->value != value) {
    curr = curr->next;
  }
  return curr;
}

void list_insert(DoublyLinkedList *list, ListNode *node) {
  node->prev = NULL;
  if (list->head != NULL)
    list->head->prev = node;
  node->next = list->head;
  list->head = node;
}

void list_delete(DoublyLinkedList *list, ListNode *node) {
  if (node != NULL) {
    if (node->prev == NULL)
      list->head = node->next;
    else
      node->prev->next = node->next;

    if (node->next != NULL)
      node->next->prev = node->prev;
  }
}

void node_print(ListNode *node) {
  if (node != NULL) {
    if (node->prev == NULL)
      printf("null");
    else
      printf("%d", node->prev->value);

    printf(" <- %d -> ", node->value);

    if (node->next == NULL)
      printf("null\n");
    else
      printf("%d\n", node->next->value);
  } else {
    printf("null");
  }
}

void list_print(DoublyLinkedList list) {
  ListNode *curr = list.head;
  while (curr != NULL) {
    node_print(curr);
    curr = curr->next;
  }
}

File: doubly_linked_list/ruby/doubly_linked_list.rb
class ListNode
  attr_accessor :prev_node, :next_node, :value

  def initialize(prev_node, next_node, value)
    @prev_node = prev_node
    @next_node = next_node
    @value = value
  end
end

class DoublyLinkedList
  attr_accessor :head

  def initialize
    @head = nil
  end

  def search(value)
    curr = @head
    while curr != nil and curr.value != value
      curr = curr.next_node
    end
    return curr
  end

  def insert(node)
    node.prev_node = nil
    if @head != nil
      @head.prev_node = node
    end
    node.next_node = @head
    @head = node
  end

  def delete(node)
    if node != nil
      if node.prev_node != nil
        node.prev_node.next_node = node.next_node
      else
        @head = node.next_node
      end

      if node.next_node != nil
        node.next_node.prev_node = node.prev_node
      end
    end
  end
end
